llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -16918,6 +16918,24 @@ bool AArch64TargetLowering::shouldFoldConstantShiftPairToMask(
     return (!C1 || !C2 || C1->getZExtValue() >= C2->getZExtValue());
   }
 
+  // We do not need to fold when this shifting used in specific load case:
+  // (ldr x, (add x, (shl (srl x, c1) 2)))
+  if (N->getOpcode() == ISD::SHL) {
+    auto C2 = dyn_cast_or_null<ConstantSDNode>(N->getOperand(1));
+    if (C2 && C2->getZExtValue() == 2) {
```

Comment:
Can we change it from C2->getZExtValue() == 2 -> C2->getZExtValue() <= 3? That should capture the common cases where the shift is free.

Reply:
Oh - should the shift amount match the loaded type? So if it is loading an i32 like the tests, it should be 2, but if it was loading i16/i64 it would need to be different shift amounts?

Reply:
Yeah sounds good. A few other types sizes: https://godbolt.org/z/GqPT3KbbT. The shift in the addressing mode needs to match the loaded size.

Reply:
https://godbolt.org/z/d6xK3xvav

add comparing with ARM64-GCC.

Reply:
and it works done!